### PR TITLE
Add a translation example

### PR DESCRIPTION
**Pull request by nrempel**
This pull request adds an example of translating words using the aligned word vectors available here: https://fasttext.cc/docs/en/aligned-vectors.html

As far as I can tell, the script is working as intended. Results are mixed but I think this is as close as we can reasonably get using word vectors.

```
> python3.9 translate.py translate hello

Loading indexes from filesystem...
Language: fr, Translation: hello
Language: it, Translation: hello
Language: it, Translation: ciao
Language: fr, Translation: bonjours
Language: fr, Translation: bonjour
Language: fr, Translation: bonsoir
Language: fr, Translation: !
Language: fr, Translation: salutations
Language: it, Translation: buongiorno
Language: it, Translation: hey
```

```
> python3.9 translate.py translate apple

Loading indexes from filesystem...
Language: it, Translation: apple
Language: fr, Translation: apple
Language: it, Translation: raspberry
Language: it, Translation: iphone
Language: it, Translation: macintosh
Language: it, Translation: ipod
Language: it, Translation: android
Language: fr, Translation: acorn
Language: it, Translation: blackberry
Language: it, Translation: marshmallow
```

### CODE REVIEW

File: instant-distance-py/examples/translations/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+Language: fr, Translation: bonjour
+Language: fr, Translation: bonsoir
+Language: fr, Translation: !
+Language: fr, Translation: salutations
+Language: it, Translation: buongiorno
+Language: it, Translation: hey
+```
+
+The translate command will download the vector data and build an index on the
+first run. Subsequent runs will be faster.
```
Comment (hartshorne) [inline instant-distance-py/examples/translations/README.md:39]: Tell them how longs this takes on your MacBook Pro

**Comment by djc** [comment]
Haven't looked at the code yet, but I feel the top-line results here are pretty shit and therefore tend to undersell the quality of the solution even though the problem in this space is that we're misusing word vectors if we're not combining them with high quality word lists. Otherwise, maybe we should pick one of the ann-benchmarks things, something like the glove dataset, if we want to show off i-d?

**Comment by nrempel** [comment]
> Haven't looked at the code yet, but I feel the top-line results here are pretty shit and therefore tend to undersell the quality of the solution even though the problem in this space is that we're misusing word vectors if we're not combining them with high quality word lists. Otherwise, maybe we should pick one of the ann-benchmarks things, something like the glove dataset, if we want to show off i-d?

The results aren't great. How do you feel about framing a bit differently to be a tool to find similar words in various languages?

I'm going to take a look at what it would take to run the ann-benchmark suite—or maybe just one of the tests.

**Comment by djc** [comment]
Sorry, "pretty shit" is maybe a harsh way of labeling it, was a bit frustrated for other reasons yesterday.

Finding "similar" words could work well, but I think lacks the wow factor that translation has, and because of the noise in the aligned word vector data, still doesn't look great: in your first example only 6/10 of the examples have the "correct" language, and in the second example exactly none of the outcomes are not-English, despite them being labeled as Italian or French.

The other approach we could take is to double down on translation but use high-quality word lists (as in https://github.com/InstantDomain/instant-segment/pull/18) to restrict the indexes (we could use dictionaries from https://github.com/wooorm/dictionaries, for example). That would potentially be a bunch more work, though.

File: instant-distance-py/examples/translations/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# Translation Example
+
+This example utilizes the pre-trained aligned word vectors made available by
+Facebook Research as part of
+[fastText](https://fasttext.cc/docs/en/aligned-vectors.html) to translate
+English words to French and Italian.
+
+## Trying it out
+
+Python 3.9 is required.
```
Comment (djc) [inline instant-distance-py/examples/translations/README.md:10]: Currently we only publish macOS wheels. Should we mention that here? Or figure out what we need to publish Linux/Windows wheels?

File: instant-distance-py/examples/translations/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Translation Example
+
+This example utilizes the pre-trained aligned word vectors made available by
```
Comment (djc) [inline instant-distance-py/examples/translations/README.md:3]: Nit: maybe replace `utilizes` with just `uses`?

File: instant-distance-py/examples/translations/translate.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+BUILT_IDX_PATH = f"./data/{'_'.join(LANGS)}.idx"
+WORD_MAP_PATH = f"./data/{'_'.join(LANGS)}.json"
+
+
+async def download_build_index():
+    """
+    This function downloads pre-trained word vectors trained on Wikipedia using fastText:
+    https://fasttext.cc/docs/en/aligned-vectors.html
+
+    The content is streamed and we take only the first 100,000 lines and drop the longtail
```
Comment (djc) [inline instant-distance-py/examples/translations/translate.py:24]: "longtail" -> "long tail"

File: instant-distance-py/examples/translations/translate.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                            bar.next()
+
+    # Build the instant-distance index and dump it out to a file with .idx suffix
+    print("Building index... (this will take a while)")
+    hnsw = instant_distance.HnswMap.build(points, values, instant_distance.Config())
+    hnsw.dump(BUILT_IDX_PATH)
+
+    # Store the mapping from string to embedding in a .json file
+    with open(WORD_MAP_PATH, "w") as f:
+        f.write(json.dumps(word_map))
```
Comment (djc) [inline instant-distance-py/examples/translations/translate.py:86]: Use `json.dump()` here?

File: instant-distance-py/examples/translations/translate.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    nearest neighbors to that point (embedding) in order to translate to other languages.
+    """
+    data_exists = os.path.isfile(BUILT_IDX_PATH) and os.path.isfile(WORD_MAP_PATH)
+    if not data_exists:
+        print("Instant Distance index not present. Building...")
+        await download_build_index()
+
+    print("Loading indexes from filesystem...")
+    with open(WORD_MAP_PATH, "r") as f:
+        word_map = json.loads(f.read())
```
Comment (djc) [inline instant-distance-py/examples/translations/translate.py:105]: Can use `json.load()` here.

File: instant-distance-py/examples/translations/translate.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+    # Get an embedding for the given word
+    try:
+        embedding = word_map[word]
+    except KeyError:
+        print(f"Word not recognized: {word}")
+        exit(1)
```
Comment (djc) [inline instant-distance-py/examples/translations/translate.py:112]: I usually prefer not to use exceptions for control flow. Why not just use `word_map.get(word)`?

File: instant-distance-py/examples/translations/translate.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                        line = await resp.content.readline()
+                        if not line:
+                            # EOF
+                            break
+
+                        # We just use the top 100k embeddings to
+                        # save on space and time
+                        elif lineno > MAX_LINES:
+                            break
+                        else:
```
Comment (djc) [inline instant-distance-py/examples/translations/translate.py:58]: Maybe skip the `else` here so we can indent a bit less in the following block?

**Comment by nrempel** [comment]
No offense taken! And thanks for the review—I agree with all of the points there.

Before I spend any time iterating on this – what do you think @hartshorne? For the purposes of the engineering article, should I take a look at improving the quality of the translations?

**Comment by hartshorne** [comment]
Let's clean this up and check it in, then let me run it past the fastText people to see if they have other ideas to improve the translations. Like we can ask them if they think filtering words based on good dictionaries is reasonable. They might argue that English words tend to get a lot of actual usage in other languages for a variety of reasons, but let's find out.

File: instant-distance-py/examples/translations/README.md (hunk 1 above)
Comment (nrempel) [reply instant-distance-py/examples/translations/README.md:39]: Decided not to include timing information as part of the script to keep it simple, and just added timing for my setup.

File: instant-distance-py/examples/translations/README.md (hunk 2 above)
Comment (nrempel) [reply instant-distance-py/examples/translations/README.md:10]: I didn't realize that. I'll update the messaging.

**Comment by nrempel** [comment]
Those changes should cover the suggestions.

**Pull request merged**